Helper/Cart.php

```diff
@@ -1917,62 +1920,102 @@ public function collectDiscounts(
                 ///////////////////////////////////////////////////////////////////////////
                 if ($discount == Discount::AMASTY_GIFTCARD && $this->discountHelper->getAmastyPayForEverything()) {
                     $giftCardCodes = $this->discountHelper->getAmastyGiftCardCodesFromTotals($totals);
-                    $amount = $this->discountHelper->getAmastyGiftCardCodesCurrentValue($giftCardCodes);
-                }
-                ///////////////////////////////////////////////////////////////////////////
-
-                ///////////////////////////////////////////////////////////////////////////
-                // Change giftcards balance as discount amount to giftcard balances to the discount amount
-                ///////////////////////////////////////////////////////////////////////////
-                if ($discount == Discount::MAGEPLAZA_GIFTCARD) {
+                    foreach($giftCardCodes as $giftCardCode) {
+                        $amount = abs($this->discountHelper->getAmastyGiftCardCodesCurrentValue(array($giftCardCode)));
+                        $roundedAmount = CurrencyUtils::toMinor($amount, $currencyCode);
+                        $discountItem = [
+                            'description' => $description . $giftCardCode,
+                            'amount'      => $roundedAmount,
+                            'discount_category' => 'giftcard',
+                            'reference'   => $giftCardCode,
+                            'discount_type'   => $this->discountHelper->getBoltDiscountType('by_fixed'),
+                        ];
+                        $discountAmount += $amount;
+                        $roundedDiscountAmount += $roundedAmount;
+                        $discounts[] = $discountItem;
+                    }
```

Comment:
Lines 1923-1936, 1942-19555 .. are very similar. Do you think it's possible to have function for them?
We can use anonymous function for example similar like we have here https://github.com/BoltApp/bolt-magento2/blob/3fc3bb8b3f710e058ab52b53167ac841c5230e1d/Helper/Cart.php#L900

Reply:
@vitaliyreznikov they are for different gift card types, and we have separate snippet for each type of discount, so I think we would better keep them separate, since the logic may change